python/ir/bindings.cpp

```diff
@@ -0,0 +1,147 @@
+#include <pybind11/functional.h>
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+#include "reffine/base/type.h"
+#include "reffine/builder/reffiner.h"
+#include "reffine/ir/expr.h"
+#include "reffine/ir/loop.h"
+#include "reffine/ir/op.h"
+#include "reffine/ir/stmt.h"
+#include "reffine/pass/printer.h"
+
+using namespace std;
+using namespace reffine;
+using namespace reffine::reffiner;
+
+namespace py = pybind11;
+
+string to_string(shared_ptr<Func> fn) { return IRPrinter::Build(fn); }
+
+#define REGISTER_CLASS(CLASS, PARENT, MODULE, NAME, ...)       \
+    py::class_<CLASS, shared_ptr<CLASS>, PARENT>(MODULE, NAME) \
+        .def(py::init<__VA_ARGS__>());
+
+PYBIND11_MODULE(ir, m)
+{
+    /* Structures related to Reffine typing */
+    py::enum_<BaseType>(m, "BaseType")
+        .value("bool", BaseType::BOOL)
+        .value("i8", BaseType::INT8)
+        .value("i16", BaseType::INT16)
+        .value("i32", BaseType::INT32)
+        .value("i64", BaseType::INT64)
+        .value("u8", BaseType::UINT8)
+        .value("u16", BaseType::UINT16)
+        .value("u32", BaseType::UINT32)
+        .value("u64", BaseType::UINT64)
+        .value("f32", BaseType::FLOAT32)
+        .value("f64", BaseType::FLOAT64)
+        .value("struct", BaseType::STRUCT)
+        .value("ptr", BaseType::PTR);
+
+    py::class_<DataType>(m, "DataType")
+        .def(py::init<BaseType, vector<DataType>, size_t>(), py::arg("btype"),
+             py::arg("dtypes") = vector<DataType>{}, py::arg("size") = 0)
+        .def("str", &DataType::str);
+
+    /* StmtNode and Derived Structures Declarations
+     */
+    py::class_<StmtNode, Stmt>(m, "stmt");
+    py::class_<ExprNode, Expr, StmtNode>(m, "expr");
+
+    /* Symbol Definition */
+    py::class_<SymNode, Sym, ExprNode>(m, "sym")
+        .def(py::init<string, DataType>())
+        .def(py::init<string, Expr>());
+
+    /* Loop */
+    REGISTER_CLASS(IsValid, ExprNode, m, "is_valid", Expr, Expr, size_t)
+    REGISTER_CLASS(SetValid, ExprNode, m, "set_valid", Expr, Expr, Expr, size_t)
+    REGISTER_CLASS(FetchDataPtr, ExprNode, m, "fetch", Expr, Expr, size_t)
+    REGISTER_CLASS(Alloc, ExprNode, m, "alloc", DataType, Expr)
+    REGISTER_CLASS(Load, ExprNode, m, "load", Expr)
+    REGISTER_CLASS(Store, StmtNode, m, "store", Expr, Expr)
+    REGISTER_CLASS(Loop, ExprNode, m, "loop", Expr)
+
+    /* Op */
+    REGISTER_CLASS(Op, ExprNode, m, "op", vector<Sym>, Expr, vector<Expr>)
+
+    /* Statements */
+    REGISTER_CLASS(Func, StmtNode, m, "func", string, Expr, vector<Sym>,
+                   SymTable)
+    REGISTER_CLASS(Stmts, StmtNode, m, "stmts", vector<Stmt>)
+    REGISTER_CLASS(IfElse, StmtNode, m, "ifelse", Expr, Stmt, Stmt)
+    REGISTER_CLASS(NoOp, StmtNode, m, "noop")
+
+    /* Misc Expressions */
+    REGISTER_CLASS(Call, ExprNode, m, "call", string, DataType, vector<Expr>)
```

Comment:
these ones should also start with underscore. to be consistent with all the other operations